plover/oslayer/xkeyboardcontrol.py

```diff
@@ -246,18 +252,905 @@ def ignore_key_events(self, key_events):
+  0x20ab: 0x20ab, #                    DongSign ₫ DONG SIGN
+  0x20ac: 0x20ac, #                    EuroSign € EURO SIGN
+}
+UCS_TO_KEYSYM = dict((ucs, keysym)
+                     for keysym, ucs
+                     in KEYSYM_TO_UCS.items())
+
+def is_latin1(code):
+    return 0x20 <= code <= 0x7e or 0xa0 <= code <= 0xff
+
+def uchr_to_keysym(char):
+    assert isinstance(char, unicode)
+    code = ord(char)
+    # Latin-1 characters: direct, 1:1 mapping.
+    if is_latin1(code):
+        return code
+    return UCS_TO_KEYSYM.get(code, code | 0x01000000)
+
+def keysym_to_string(keysym):
+    # Latin-1 characters: direct, 1:1 mapping.
+    if is_latin1(keysym):
+        code = keysym
+    elif (keysym & 0xff000000) == 0x01000000:
+        code = keysym & 0x00ffffff
+    else:
+        code = KEYSYM_TO_UCS.get(keysym)
+        if code is None:
+            keysym_str = XK.keysym_to_string(keysym)
+            if keysym_str is None:
+                keysym_str = u''
+            for c in keysym_str:
+                if c not in string.printable:
+                    keysym_str = u''
+                    break
+            return keysym_str
+    return unichr(code)
+
+
 class KeyboardEmulation(object):
     """Emulate keyboard events."""
 
+    class Mapping(object):
+
+        def __init__(self, keycode, modifiers, keysym, custom_mapping=None):
+            self.keycode = keycode
+            self.modifiers = modifiers
+            self.keysym = keysym
+            self.custom_mapping = custom_mapping
+
+        def __str__(self):
+            return u'%u:%x=%x[%s]%s' % (
+                self.keycode, self.modifiers,
+                self.keysym, keysym_to_string(self.keysym),
+                '' if self.custom_mapping is None else '*',
+            )
+
+    # We can use the first 2 entry of a X11 mapping:
+    # keycode and keycode+shift. The 3rd entry is a
+    # special keysym to mark the mapping.
+    CUSTOM_MAPPING_LENGTH = 3
+    # Special keysym to mark custom keyboard mappings.
+    PLOVER_MAPPING_KEYSYM = 0x01ffffff
+
     def __init__(self):
         """Prepare to emulate keyboard events."""
         self.display = display.Display()
-        self.modifier_mapping = self.display.get_modifier_mapping()
-        # Determine the backspace keycode.
-        backspace_keysym = XK.string_to_keysym('BackSpace')
-        self.backspace_keycode, mods = self._keysym_to_keycode_and_modifiers(
-                                                backspace_keysym)
         self.time = 0
+        self._update_keymap()
+
+    def _update_keymap(self):
+        '''Analyse keymap, build a mapping of keysym to (keycode + modifiers),
+        and find unused keycodes that can be used for unmapped keysyms.
+        '''
+        self.keymap = {}
+        self.custom_mappings_queue = []
+        # Analyse X11 keymap.
+        keycode = self.display.display.info.min_keycode
+        nb_keycodes = self.display.display.info.max_keycode - keycode + 1
```

Comment:
That would normally be clear, but this case is special, because we're dealing with keycodes. What's confusing with any identifier name using just "num" here is that it could be abbreviating "numeric". Presumably there are keycodes that correspond to the number keys, and those would be either "number keycodes" or "numeric keycodes", and then I'm confused again. The code clears it up, but it's a confusion we can avoid, either by spelling out `number_of_keycodes` or flipping it around to `keycode_count`.